src/pykeen/utils.py

```diff
@@ -1095,3 +1100,118 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
     import doctest
 
     doctest.testmod()
+
+
+def product_normalise(input_tensor: torch.FloatTensor) -> torch.FloatTensor:
+    r"""Normalise the input tensor along its embedding dimension so that the geometric mean is 1.
+
+    :param input_tensor: An input tensor with final dimension $d$.
+
+    :returns: An output tensor whose last order is normalized to have a geometric mean of 1.
+    """
+    step1_tensor = torch.abs(input_tensor)  # Compute absolute value of all entries
+    step2_tensor = step1_tensor.clamp(min=SANITY_EPSILON)  # Prevent zero values by adding a sanity epsilon
+    log_norm_tensor = torch.log(step2_tensor)  # Compute the log prior to computing the geom. mean
+    step3_tensor = torch.mean(log_norm_tensor, dim=-1, keepdim=True)
+    norm_volume = torch.exp(step3_tensor)
+    pre_norm_out = input_tensor / norm_volume
+    return pre_norm_out
```

Comment:
```suggestion
    return input_tensor / input_tensor.abs()\
       .clamp(min=SANITY_EPSILON)\
       .log()\
       .mean(dim=1, keepdim=True)\
       .exp()\
       .clamp(min=SANITY_EPSILON)
```

not super beautiful, but does not keep references to intermediate tensors, and thus allows earlier garbage collection / memory re-use.

I also added clamping before division - I think this should not be necessary (since `exp() > 0`), but just to be sure for numeric instability.

Reply:
All good! So long as it's equivalent :)

Reply:
Isn't `input_tensor / input_tensor.abs()` also the same as taking the norm with `p=1`? Then you can achieve maximum chainage

Reply:
> Isn't input_tensor / input_tensor.abs() also the same as taking the norm with p=1? Then you can achieve maximum chainage

no, the full right hand side is evaluated before the division. I'll add a bracket to make this clearer :sweat_smile:

Reply:
just let black do what it wants and add a code comment to prevent future non-careful readers like me ;)

Reply:
Added in d7264a4a